### DISCUSSION TITLE
How do you speed up the sample rate in DMP?

### DISCUSSION THREAD

**Discussion by schosdas** [jrowberg/i2cdevlib#668] [Q&A]
I want to print more than 500 data per second in the DMP example. 
I heard that there is a way to modify the sample rate. 
After mpu initialization, I tried "mpu.setRate(1);", but the output speed still seems to be there. How can I print out more data?

```
void setup() {
  // join I2C bus (I2Cdev library doesn't do this automatically)
#if I2CDEV_IMPLEMENTATION == I2CDEV_ARDUINO_WIRE
  Wire.begin();
  Wire.setClock(400000); // 400kHz I2C clock. Comment this line if having compilation difficulties
  Wire.setWireTimeout(3000, true); // for stop mpu error
#elif I2CDEV_IMPLEMENTATION == I2CDEV_BUILTIN_FASTWIRE
  Fastwire::setup(400, true);
#endif

  // initialize serial communication
  // (115200 chosen because it is required for Teapot Demo output, but it's
  // really up to you depending on your project)
  Serial.begin(115200);
  while (!Serial); // wait for Leonardo enumeration, others continue immediately

  // NOTE: 8MHz or slower host processors, like the Teensy @ 3.3V or Arduino
  // Pro Mini running at 3.3V, cannot handle this baud rate reliably due to
  // the baud timing being too misaligned with processor ticks. You must use
  // 38400 or slower in these cases, or use some kind of external separate
  // crystal solution for the UART timer.

  // initialize device
  Serial.println(F("Initializing I2C devices..."));
  mpu.initialize();
  pinMode(INTERRUPT_PIN, INPUT);

  // verify connection
  Serial.println(F("Testing device connections..."));
  Serial.println(mpu.testConnection() ? F("MPU6050 connection successful") : F("MPU6050 connection failed"));

  // wait for ready
  Serial.println(F("\nSend any character to begin DMP programming and demo: "));
  while (Serial.available() && Serial.read()); // empty buffer
  while (!Serial.available());                 // wait for data
  while (Serial.available() && Serial.read()); // empty buffer again

  // load and configure the DMP
  Serial.println(F("Initializing DMP..."));
  devStatus = mpu.dmpInitialize();

  // supply your own gyro offsets here, scaled for min sensitivity
  mpu.setXGyroOffset(220);
  mpu.setYGyroOffset(76);
  mpu.setZGyroOffset(-85);
  mpu.setZAccelOffset(1788); // 1688 factory default for my test chip

  // make sure it worked (returns 0 if so)
  if (devStatus == 0) {
    // Calibration Time: generate offsets and calibrate our MPU6050
    mpu.CalibrateAccel(6);
    mpu.CalibrateGyro(6);
    mpu.PrintActiveOffsets();
    // turn on the DMP, now that it's ready
    Serial.println(F("Enabling DMP..."));
    mpu.setDMPEnabled(true);

    // enable Arduino interrupt detection
    Serial.print(F("Enabling interrupt detection (Arduino external interrupt "));
    Serial.print(digitalPinToInterrupt(INTERRUPT_PIN));
    Serial.println(F(")..."));
    attachInterrupt(digitalPinToInterrupt(INTERRUPT_PIN), dmpDataReady, RISING);
    mpuIntStatus = mpu.getIntStatus();

    // set our DMP Ready flag so the main loop() function knows it's okay to use it
    Serial.println(F("DMP ready! Waiting for first interrupt..."));
    dmpReady = true;

    // get expected DMP packet size for later comparison
    packetSize = mpu.dmpGetFIFOPacketSize();

    //test
    mpu.setRate(1); // 1khz / (1 + 5) = 166 Hz

  } else {
    // ERROR!
    // 1 = initial memory load failed
    // 2 = DMP configuration updates failed
    // (if it's going to break, usually the code will be 1)
    Serial.print(F("DMP Initialization failed (code "));
    Serial.print(devStatus);
    Serial.println(F(")"));
  }

  // configure LED for output
  pinMode(LED_PIN, OUTPUT);
}
```

Thank you.

**Comment by jrowberg** [comment]
Unfortunately, the short answer is that you can't do this. The DMP has a maximum output rate of 200 Hz, and this tends to be rather noisy. 100 Hz is more realistic for cleaner, smoother data. At least, this is the case as far as I know. One reason is that the DMP fusion algorithms are computationally intensive, and the small onboard processor in the IMU can't perform the math faster than this. Another reason is that simply gathering the raw data takes time, and the faster the IMU goes, the less filtering it can do and the less time it has to perform fusion computations (which are hard to do).

If you want very fast fused data, you will probably need to select a fast 6-axis IMU (likely with a 3MHz+ SPI interface) and pair it with a fast Cortex-M3/M4 MCU, with the MCU doing the fusion. This is not anywhere near as simple, of course, but that's the cost of extremely fast fused orientation data.

**Reply by schosdas** [reply]
Hello,
So, can I use 2 mpu6050 (0x68,0x69) to output raw data rather than DMP at least 500 sample rates per second?
I'm going to use two or three sensors.

**Reply by jrowberg** [reply]
The datasheet:

https://invensense.tdk.com/wp-content/uploads/2015/02/MPU-6000-Datasheet1.pdf#page=12

...indicates that 500 Hz raw data is possible from both the accelerometer and gyroscope, so what you want to do may be possible as long as the I2C bus can keep up. The MPU-6000 supports SPI and affords a faster transfer mechanism, if you need it.

Doing some quick math: if you do a single burst read across accel+temp+gyro registers (which is going to be faster than two reads of just accel and gyro due to transaction overhead), that's 17 bytes transferred total (`DevAddr + RegAddr | DevAddr + 14x read bytes`), roughly 10 bits per byte = 140 bits * 500 = 70 kbit/sec, which should fit within a 400 kHz I2C bus limit even with processing overhead. However, it might not leave very much time for fusing the data or doing whatever else you need to with it. Certainly not an an Arduino Uno, at least. One of the newer Teensy boards could probably handle it. If you can somehow use DMA for the I2C transfers, even better.
